rnn_speech_recognition/pytorch/README.md

```diff
@@ -1,44 +1,192 @@
+
 ### Training data order
-In what order is the training data traversed?
+To reduce data padding in minibatches, data bucketing is applied.
+The algorithm is implemented here:
+[link](https://github.com/mlcommons/training/blob/2126999a1ffff542064bb3208650a1e673920dcf/rnn_speech_recognition/pytorch/common/data/dali/sampler.py#L65-L105)
+and can be described as follows:
+1. drop samples longer than a given threshold;
+1. sort data by audio length;
+2. split data into 6 equally sized buckets;
+3. for every epochs:
+    1. shuffle data in each bucket;
+    2. as long as all samples are not divisible by global batch size, remove random element from random bucket;
+    3. concatenate all buckets;
+    4. split samples into minibatches;
+    5. shuffle minibatches in the epoch.
+
 ### Test data order
-In what order is the test data traversed?
-### Simulation environment (RL models only)
-Describe simulation environment briefly, if applicable. 
+Test data order is the same as in the dataset.
+
 # 4. Model
 ### Publication/Attribution
-Cite paper describing model plus any additional attribution requested by code authors 
+To the best of our knowledge, there is no single publication describing RNN-T training on LibriSpeech,
+or another publicly available dataset of reasonable size. For that reason, the reference will be a
+collection of solutions from several works. It is based on the following articles:
+* Graves 2012 - an invention of RNN-Transducer: https://arxiv.org/abs/1211.3711
+* Rao 2018 - time reduction in the acoustic model, internal dataset: https://arxiv.org/abs/1801.00841
+* Zhang 2020 - Transformer-transducer publication. It includes bi-directional LSTM RNN-T result on LibriSpeech: https://arxiv.org/abs/2002.02562
+* Park 2019 - adaptive spec augment, internal dataset: https://arxiv.org/abs/1912.05533
+* Guo 2020 - RNN-T trained with vanilla LSTM, internal dataset: https://arxiv.org/abs/2007.13802
+
 ### List of layers 
-Brief summary of structure of model
+Model structure is described in the following picture:
+![model layers structure](./rnnt_layers.svg "RNN-T model structure")
+
 ### Weight and bias initialization
-How are weights and biases initialized
+* In all fully connected layers, weights and biases are initialized as defined in the [Pytorch 1.7.0 torch.nn.Linear documentation](https://pytorch.org/docs/1.7.0/generated/torch.nn.Linear.html#torch.nn.Linear).
+* In the embeding layer, weights are initialized as defined in the [Pytorch 1.7.0 torch.nn.Embeding documentation](https://pytorch.org/docs/1.7.0/generated/torch.nn.Embedding.html#torch.nn.Embedding).
+* In all LSTM layers:
+    * weights and biases are initialized as defined in the [Pytorch 1.7.0 torch.nn.LSTM documentation](https://pytorch.org/docs/1.7.0/generated/torch.nn.LSTM.html#torch.nn.LSTM),
+    * then the weights and bias values are divided by two,
+    * forget gate biases are set to 1.
```

Comment:
might want to include a link to the code here as its a confusing detail.
[forget-gate-bias](https://github.com/mwawrzos/training/blob/rnnt/rnn_speech_recognition/pytorch/common/rnn.py#L65)

Reply:
When I was adding links to this section I noticed, that biases are halved after setting to 1. So these are set to 0.5 by mistake.
The model converges, so I don't think it needs a change. I updated the readme to highlight this fact.
I like the idea of giving links to sources, so I added in 4e75866d2b1e54b9351b0f6c2d243266b95ed3fc links to other places too.